lib/http3/common.c

```diff
@@ -713,37 +720,73 @@ void h2o_http3_read_socket(h2o_http3_ctx_t *ctx, h2o_socket_t *sock)
-                  dgrams[dgram_index - 1].ttl != dgrams[dgram_index].ttl)) {
-                process_packets(ctx, &dgrams[dgram_index - 1].destaddr, &dgrams[dgram_index - 1].srcaddr,
-                                dgrams[dgram_index - 1].ttl, packets, packet_index);
-                packet_index = 0;
+        dgram_index = 0;
+        while (dgram_index < num_dgrams) {
+            int has_decoded = 0; /* indicates if a decoded packet belonging to a different connection is stored at
+                                  * `packets[packet_index]` */
+            /* dispatch packets in `packets`, if the datagram at dgram_index is from a different path */
+            if (packet_index != 0) {
+                assert(dgram_index != 0);
+                /* check source address */
+                if (h2o_socket_compare_address(&dgrams[dgram_index - 1].srcaddr.sa, &dgrams[dgram_index].srcaddr.sa, 1) != 0)
+                    goto ProcessPackets;
+                /* check destination address, if available */
+                if (dgrams[dgram_index - 1].destaddr.sa.sa_family == AF_UNSPEC &&
+                    dgrams[dgram_index].destaddr.sa.sa_family == AF_UNSPEC) {
+                    /* ok */
+                } else if (h2o_socket_compare_address(&dgrams[dgram_index - 1].destaddr.sa, &dgrams[dgram_index].destaddr.sa, 1) ==
+                           0) {
+                    /* ok */
+                } else {
+                    goto ProcessPackets;
+                }
+                /* check TTL */
+                if (dgrams[dgram_index - 1].ttl != dgrams[dgram_index].ttl)
+                    goto ProcessPackets;
+            }
+            /* decode the first packet */
+            size_t payload_off = 0;
+            if (quicly_decode_packet(ctx->quic, packets + packet_index, dgrams[dgram_index].vec.iov_base,
+                                     dgrams[dgram_index].vec.iov_len, &payload_off) == SIZE_MAX) {
+                ++dgram_index;
+                goto ProcessPackets;
             }
-            size_t off = 0;
-            while (off < dgrams[dgram_index].vec.iov_len) {
+            /* dispatch packets in `packets` if the DCID is different, setting the `has_encoded` flag */
+            if (packet_index != 0) {
+                const ptls_iovec_t *prev_dcid = &packets[packet_index - 1].cid.dest.encrypted,
+                                   *cur_dcid = &packets[packet_index].cid.dest.encrypted;
+                if (!(prev_dcid->len == cur_dcid->len && memcmp(prev_dcid->base, cur_dcid->base, prev_dcid->len) == 0)) {
+                    has_decoded = 1;
+                    ++dgram_index;
+                    goto ProcessPackets;
+                }
+            } else {
+                ++packet_index;
+            }
+            /* add rest of the packets */
+            while (payload_off < dgrams[dgram_index].vec.iov_len && packet_index < PTLS_ELEMENTSOF(packets)) {
```

Comment:
A peer MUST NOT coalesce packets from different connections. While not checking here will work fine, is it worth catching it and calling a PROTOCOL_VIOLATION? Certainly not necessary for this PR.

Reply:
We cannot call it a PROTOCOL_VIOLATION, because "coalescing" is not authenticated. We might want to limit the maximum number of QUIC packets that we extract from one datagram though.